streams/src/main/java/org/apache/kafka/streams/state/internals/BufferValue.java

```diff
@@ -16,83 +16,162 @@
+
+        final byte[] priorValue = extractValue(buffer);
+
+        final byte[] oldValue;
+        final int oldValueLength = buffer.getInt();
+        if (oldValueLength == NULL_VALUE_SENTINEL) {
+            oldValue = null;
+        } else if (oldValueLength == OLD_PREV_DUPLICATE_VALUE_SENTINEL) {
+            oldValue = priorValue;
+        } else {
+            oldValue = new byte[oldValueLength];
+            buffer.get(oldValue);
+        }
+
+        final byte[] newValue = extractValue(buffer);
+
+        return new BufferValue(priorValue, oldValue, newValue, context);
+    }
 
-        final int priorValueLength = buffer.getInt();
-        if (priorValueLength == -1) {
-            return new BufferValue(record, null);
+    private static byte[] extractValue(final ByteBuffer buffer) {
+        final int valueLength = buffer.getInt();
+        if (valueLength == NULL_VALUE_SENTINEL) {
+            return null;
         } else {
-            final byte[] priorValue = new byte[priorValueLength];
-            buffer.get(priorValue);
-            return new BufferValue(record, priorValue);
+            final byte[] value = new byte[valueLength];
+            buffer.get(value);
+            return value;
         }
     }
 
     ByteBuffer serialize(final int endPadding) {
 
-        final int sizeOfPriorValueLength = Integer.BYTES;
+        final int sizeOfValueLength = Integer.BYTES;
+
         final int sizeOfPriorValue = priorValue == null ? 0 : priorValue.length;
+        final int sizeOfOldValue = oldValue == null || priorValue == oldValue ? 0 : oldValue.length;
+        final int sizeOfNewValue = newValue == null ? 0 : newValue.length;
+
+        final byte[] serializedContext = recordContext.serialize();
+
+        final ByteBuffer buffer = ByteBuffer.allocate(
+            serializedContext.length
+                + sizeOfValueLength + sizeOfPriorValue
+                + sizeOfValueLength + sizeOfOldValue
+                + sizeOfValueLength + sizeOfNewValue
+                + endPadding
+        );
 
-        final ByteBuffer buffer = record.serialize(sizeOfPriorValueLength + sizeOfPriorValue + endPadding);
+        buffer.put(serializedContext);
 
-        if (priorValue == null) {
-            buffer.putInt(-1);
+        addValue(buffer, priorValue);
+
+        if (oldValue == null) {
+            buffer.putInt(NULL_VALUE_SENTINEL);
+        } else if (priorValue == oldValue) {
+            buffer.putInt(OLD_PREV_DUPLICATE_VALUE_SENTINEL);
         } else {
-            buffer.putInt(priorValue.length);
-            buffer.put(priorValue);
+            buffer.putInt(sizeOfOldValue);
+            buffer.put(oldValue);
         }
 
+        addValue(buffer, newValue);
+
         return buffer;
     }
 
-    long sizeBytes() {
-        return (priorValue == null ? 0 : priorValue.length) + record.sizeBytes();
+    private static void addValue(final ByteBuffer buffer, final byte[] newValue) {
```

Comment:
It's a little confusing that this is named newValue, but is sometimes actually priorValue

Reply:
oh, refactoring mistake. I'll fix it. Thanks!